### DISCUSSION TITLE
hen egg problem - need help - modify soap request url for each query

### DISCUSSION THREAD

**Discussion by michabbb** [phpro/soap-client#435] [Q&A]
hi there,
i am trying to use this package for the ebay soap API.
the problem here is, that for each request, the request URL needs to be extended with some infos,
so the target request URL in the WSDL is useless because with each request, the request name for instance
has to be part of the URL.

after digging in the all the issues and documentation, I am aware, that there are plugins, middleware, and stuff like that,
I didn´t fully understand till now. so I - maybe - did understand, that I can do something like this:

```php
$engine = DefaultEngineFactory::create(
            ExtSoapOptions::defaults($wsdl, [])
                ->withClassMap(eBayTradingClassmap::getCollection()),
                    Psr18Transport::createForClient(new PluginClient(Psr18ClientDiscovery::find(),...$middleware))
        );
```

don´t know if that construct is correct. but If i am not wrong, a "middleware" should be able to modify something here "on-the-fly", i hope. 

so in my case, the base URL is: `https://api.ebay.com/wsapi` but when i use this soap-client with this function:

`$response = $client->geteBayOfficialTime(new GeteBayOfficialTimeRequestType());`

the request URL has to be: `https://api.ebay.com/wsapi?callname=' . $method . '&siteid=' . $siteId . '&appid=' . $appid . '&version=' . $version . '&routing=new'`

if the `__location` would be fixed, i could set it with `ExtSoapOptions::defaults`, but in my case, it has to change for each request
i do, but i have no clue how to achieve this ;(

and i would be very thankful if someone could point me to the right direction, how i can manage this somehow.

thanks !!!

**Comment by veewee** [comment]
Hello,

Yes, building a custom HTTP middleware that dynamically changes the request would be the way to go here.
More detailed information on building your own:
https://docs.php-http.org/en/latest/plugins/build-your-own.html

I imagine siteId, appId and version are hardcoded?
The method name can be fetched from the request:
https://github.com/php-soap/psr18-transport/blob/main/src/HttpBinding/SoapActionDetector.php

**Reply by michabbb** [reply]
thanks a lot for your quick feedback!
when debugging the request, "SOAPAction" in my header is always empty, which is - in my case - not a problem.
so I don´t know how to use that. but I was able to manage it this way:

maybe not the most elegant way, but working 😏 

```php
$queryDefaultsPlugin = new QueryAddEbayParamters([
                                                             'siteid'  => $siteId,
                                                             'appid'   => $appid,
                                                             'version' => $version,
                                                             'routing' => 'new'
                                                         ]);

// ---------------------------------------------------------------------------------------------

final class QueryAddEbayParamters implements Plugin
{
    private array $queryParams = [];

    public function __construct(array $queryParams)
    {
        $this->queryParams = $queryParams;
    }

public function handleRequest(RequestInterface $request, callable $next, callable $first): Promise
    {
        $uri = $request->getUri();

        $request->getBody()->rewind();
        $body = $request->getBody()->getContents();
        $request->getBody()->rewind();

        /** @noinspection RegExpUnnecessaryNonCapturingGroup */
        $re = '/<\w+:(\w+)Request(?:\/?)>/';

        preg_match_all($re, $body, $matches, PREG_SET_ORDER, 0);

        if (!count($matches)) {
            throw new \RuntimeException('unable to find SOAP Request Callname');
        }
        if (count($matches[0])!==2) {
            throw new \RuntimeException('unable to find SOAP Request Callname');
        }
        $method = $matches[0][1];

        $this->queryParams['callname'] = $method;

        parse_str($uri->getQuery(), $query);
        $query += $this->queryParams;

        $request = $request->withUri(
            $uri->withQuery(http_build_query($query))
        );

        return $next($request);
    }
}
```
